jetson/ra_kinematics/motion_planner.cpp

```diff
@@ -0,0 +1,340 @@
+#include "motion_planner.hpp"
+
+#include <random>
+#include <queue>
+#include <time.h>
+
+
+MotionPlanner::MotionPlanner(const ArmState &robot, KinematicsSolver &solver_in) :
+        solver(solver_in) { 
+    step_limits.reserve(6);
+
+    // add limits for each joint to joint_limits, after converting to degrees
+    for (size_t i = 0; i < 6; ++i) {
+      vector<double> limits = robot.get_joint_limits(i);
+
+      joint_limits.push_back(limits);
+      step_limits.push_back(robot.get_joint_max_speed(i) / 50.0); //TODO scale if path planning takes too long
+    }
+
+    //time_t timer;
+    std::default_random_engine eng(clock());
+}
+
+Vector6d MotionPlanner::sample(Vector6d start, const ArmState &robot) {
+    Vector6d z_rand;
+
+    for (size_t i = 0; i < joint_limits.size(); ++i) {
+        if (robot.get_joint_locked(i)) {
+            // if joint is locked, z_rand matches start
+            z_rand(i) = start(i);
+        }
+        else { 
+            // create distribution of angles between limits and choose an angle
+            std::uniform_real_distribution<double> distr(joint_limits[i][0], joint_limits[i][1]);
+            z_rand(i) = distr(eng);
+        }
+    }
+
+    return z_rand;
+}
+
+MotionPlanner::Node* MotionPlanner::nearest(MotionPlanner::Node* tree_root, const Vector6d &rand) {
+    queue<Node*> q;
+    q.push(tree_root);
+
+    double min_dist = numeric_limits<double>::max();
+    Node* min_node = nullptr;
+
+    // Run breadth-first search for nearest node in tree
+    while (!q.empty()) {
+        Node* node = q.front();
+        q.pop();
+
+        // TODO consider taking into account cost
+        double dist = (node->config - rand).norm();
```

Comment:
is there a cost function right now?

Reply:
Idk I'm a bit confused by that comment, it can be deleted. Most of the motion planning has worked great and the bulk of that hasn't been touched too much since the original transcription from python